serialization/src/main/kotlin/io/kinference/protobuf/message/TensorProto.kt

```diff
@@ -0,0 +1,210 @@
+import java.nio.ByteBuffer
+import java.nio.ByteOrder
+
+class TensorProto(
+    var dims: IntArray = IntArray(0),
+    var dataType: DataType? = null,
+    var segment: Segment? = null,
+    val stringData: MutableList<ByteString> = ArrayList(),
+    var name: String? = null,
+    val externalData: MutableList<StringStringEntryProto> = ArrayList(),
+    var dataLocation: DataLocation? = null,
+) {
+    private var _tiledData: TiledArrayContainer = TiledArrayContainer()
+
+    val tiledData: Any?
+        get() = _tiledData.get(dims)
+
+    fun isTiled(): Boolean = _tiledData.hasData()
+    fun isString(): Boolean = stringData.isNotEmpty()
+
+    companion object {
+        private val int32AvailableTypes = setOf(DataType.BOOL, DataType.INT8, DataType.UINT8, DataType.INT16, DataType.UINT16)
+
+        fun decode(byteArray: ByteArray): TensorProto {
+            val buffer = Buffer().write(byteArray)
+            return decode(ProtobufReader(buffer))
+        }
+
+        fun decode(reader: ProtobufReader): TensorProto {
+            val proto = TensorProto()
+            var rawData: ByteString? = null
+            reader.forEachTag { tag ->
+                when (tag) {
+                    1 -> proto.dims = reader.readLongArray(tag).toIntArray()
+                    2 -> proto.dataType = DataType.fromValue(reader.readInt())
+                    3 -> proto.segment = Segment.decode(reader)
+                    4 -> reader.readFloatTiledArray(tag, proto.dims, proto._tiledData)
+                    5 -> reader.readIntTiledArray(tag, proto.dims, proto._tiledData)
+                    6 -> proto.stringData.add(reader.readBytes())
+                    7 -> reader.readLongTiledArray(tag, proto.dims, proto._tiledData)
+                    8 -> proto.name = reader.readString()
+                    9 -> rawData = reader.readBytes()
+                    10 -> reader.readDoubleTiledArray(tag, proto.dims, proto._tiledData)
+                    11 -> reader.readULongTiledArray(tag, proto.dims, proto._tiledData)
+                    12 -> reader.readString() // skip docstring
+                    13 -> proto.externalData.add(StringStringEntryProto.decode(reader))
+                    14 -> try {
+                        proto.dataLocation = reader.readValue(DataLocation.ADAPTER)
+                    } catch (e: ProtoAdapter.EnumConstantNotFoundException) {
+                        reader.addUnknownField(tag, FieldEncoding.VARINT, e.value.toLong())
+                    }
+                    else -> reader.readUnknownField(tag)
+                }
+            }
+            if (rawData != null || !proto.hasData()) parseRaw(rawData, proto)
+            proto.checkTiledData()
+            return proto
+        }
+
+        private fun TensorProto.hasData() = _tiledData.hasData() || stringData.isNotEmpty() || externalData.isNotEmpty()
+
+        // convert data stored as int32 to the specified type
+        private fun TensorProto.checkTiledData() {
```

Comment:
Can we read int and immediately convert it to the desired type?

Reply:
The data is not ordered, so it's more safe to read the full message and then process the data when we are guaranteed to have both the type and the data